src/tool/xmeta/xmeta_emit.cpp

```diff
@@ -0,0 +1,674 @@
+        signature_blob value_signature;
+        value_signature.add_signature(FieldSig{ TypeSig{underlying_type} });
+        mdFieldDef field_token;
+        check_hresult(m_metadata_emitter->DefineField(token_enum_type_def,
+            L"value__",
+            enum_value_flag,
+            value_signature.data(),
+            value_signature.size(),
+            ELEMENT_TYPE_END,
+            nullptr,
+            0,
+            &field_token));
+
+        auto iter = type_references.find(type_name);
+        if (iter == type_references.end())
+        {
+            throw_invalid("Failed to find TypeRef for: " + type_name);
+        }
+        TypeRef const& enum_type_ref = iter->second;
+
+        static constexpr DWORD enumerator_flag = fdHasDefault | fdLiteral | fdStatic | fdPublic;
+        signature_blob enumerator_signature;
+        enumerator_signature.add_signature(FieldSig{ TypeSig{ElementType::ValueType, enum_type_ref.coded_index<TypeDefOrRef>()} });
+
+        for (enum_member const& enum_member : model->get_members())
+        {
+            call(enum_member.get_resolved_value(), [&](auto const& val)
+                {
+                    using val_type = std::decay_t<decltype(val)>;
+                    static_assert(std::is_integral_v<val_type>);
+                    auto const& name = s2ws(enum_member.get_id());
+                    check_hresult(m_metadata_emitter->DefineField(token_enum_type_def,
+                        name.c_str(),
+                        enumerator_flag,
+                        enumerator_signature.data(),
+                        enumerator_signature.size(),
+                        static_cast<DWORD>(underlying_type),
+                        &val,
+                        static_cast<ULONG>(sizeof(val_type)),
+                        &field_token));
+                });
+        }
+    }
+    
+    void xmeta_emit::listen_delegate_model(delegate_model const& model) 
+    {
+        mdTypeDef implements[] = { mdTokenNil };
+        auto token_delegate_type_def = define_type_def(model.get_id(), delegate_type_flag, token_delegate, implements);
+
+        /* Define return value */
+        // To be used for attributes later
+        mdParamDef token_return = define_return(model.get_return_type(), token_delegate_type_def);
+
+        /* Define formal parameters */
+        int index = 1;
+        for (auto const& val : model.get_formal_parameters())
+        {
+            define_parameters(val, token_delegate_type_def, index);
+            index++;
+        }
+    }
+
+    mdParamDef xmeta_emit::define_return(std::optional<type_ref> const& retun_type, mdTypeDef const& type_def)
+    {
+        /* Define return value */
+        mdParamDef token_param_return;
+        std::string return_name = type_semantics_to_string(retun_type->get_semantic());
+        check_hresult(m_metadata_emitter->DefineParam(
+            type_def,
```

Comment:
This can't be right. The first arg to DefineParam needs to be a mdMethodDef. There's a case to be made for primarily using xlang's reader types for the type safety to avoid this sort of issue, and to convert to/from the tokens as close as possible to IMetaDataEmit.

Reply:
Also, was this meant to go into this PR? I thought this was just about the enums, not methods.

Reply:
Oh yes this is all going to be changed and not meant for it to be the PR

Reply:
But I think the define_return method will stay the same except for the name of it. It' s just the way delegate will use it will change.